src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+    private static final String  microarrayTypesBaseQuery = "type=listAttributes&mart=ENSEMBL_MART_ENSEMBL&virtualSchema=default&dataset=BIOMART_SPECIES_NAME_gene_ensembl&interface=default&attributePage=feature_page&attributeGroup=external&attributeCollection=microarray";
+
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws Exception
+     */
+    protected void downloadData() throws Exception {
+
+        // TODO: Add unit tests
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String biomartSpeciesName : getBiomartSpeciesNames()) {
+            logger.info("Retrieving Biomart files for " + biomartSpeciesName);
+            // Create directory where Biomart files will be stored.
+            new File(this.destination).mkdirs();
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(biomartSpeciesName), 0);
+            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray);
+            retrieveAndStoreMicroarrayData(biomartSpeciesName, getBiomartXMLPath(), microarrayTypes);
+
+            // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from Biomart
+            // and write it to a file (eg: hsapiens_uniprot).
+            logger.info("Retrieving UniProt data");
+            retrieveAndStoreUniprotData(biomartSpeciesName, getBiomartXMLPath());
+
+            logger.info("Completed Biomart data retrieval for " + biomartSpeciesName);
+        }
+    }
+
+    /**
+     * Query Biomart for both Uniprot mapping types (curated -- SwissProt, generated -- TrEMBL).
+     * @param biomartSpeciesName
+     * @throws Exception
+     */
+    private void retrieveAndStoreUniprotData(String biomartSpeciesName, String biomartXMLPath) throws IOException {
+        Set<String> uniprotIdLines = new HashSet<>();
+        try {
+            for (String uniprotQueryId : Arrays.asList("uniprotswissprot", "uniprotsptrembl")) {
+                // If proper response, write data to file.
+                Set<String> uniprotResponseLines = queryBiomart(getBiomartIdentifierQuery(biomartXMLPath, biomartSpeciesName, uniprotQueryId), 0);
+                for (String uniprotResponseLine : uniprotResponseLines) {
+                    List<String> tabSplit = Arrays.asList(uniprotResponseLine.split("\t"));
+                    if (tabSplit.size() == 4) {
+                        uniprotIdLines.add(uniprotResponseLine + "\n");
+                    }
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```

Comment:
Can this try block be wrapped only around the line(s) that cause the exception and the reason for the exception logged (i.e. `logger.error("Reason for exception, e);`)?

Reply:
If the try-catch is moved deeper, it won't be equivalent to the current code.
In this current state, an exception would break both loops. If the try-catch is inside one of the loops, an exception would only break the inner loop.
I would assume that @cookersjs did it this way for a reason.